cmds/contest-generator/main.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"path"
+	"strings"
+	"text/template"
+
+	flag "github.com/spf13/pflag"
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	flagTemplate = flag.StringP("template", "t", "contest.go.template", "Template file name")
+	flagOutfile  = flag.StringP("outfile", "o", "", "Output file path")
+	flagFromFile = flag.StringP("from", "f", "core-plugins.yml", "File name to get the plugin list from, in YAML format")
+)
+
+func usage() {
+	fmt.Fprintf(flag.CommandLine.Output(),
+		"%s [--template tpl] [--outfile out] [--from pluginlist]\n",
+		os.Args[0],
+	)
+	flag.PrintDefaults()
+	fmt.Fprintln(flag.CommandLine.Output(), `  plugin
+	Plugin name to include, as go import path.
+	Short names can be used for builtin contest plugins (eg "sshcmd")`)
+}
+
+var funcMap = map[string]interface{}{
+	"base": path.Base,
+}
+
+// Config is a configuration object mapped to the plugin configuration file.
+type Config struct {
+	TargetManagers []string
+	TestFetchers   []string
+	TestSteps      []string
+	Reporters      []string
+}
+
+func (t *Config) String() string {
+	ret := fmt.Sprintf("TargetManagers \n    %s\n", strings.Join(t.TargetManagers, "\n    "))
+	ret += fmt.Sprintf("TestFetchers\n    %s\n", strings.Join(t.TestFetchers, "\n    "))
+	ret += fmt.Sprintf("TestSteps\n    %s\n", strings.Join(t.TestSteps, "\n    "))
+	ret += fmt.Sprintf("Reporters\n    %s\n", strings.Join(t.Reporters, "\n    "))
+	return ret
+}
+
+func parseConfig(r io.Reader) (*Config, error) {
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file: %w", err)
+	}
+	var cfg Config
+	if err := yaml.Unmarshal(data, &cfg); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal YAML: %w", err)
+	}
+	if len(cfg.TargetManagers) < 1 {
+		return nil, fmt.Errorf("want at least 1 targetmanager, got %d", len(cfg.TargetManagers))
+	}
+	if len(cfg.TargetManagers) < 1 {
```

Comment:
all 4 conditions in this block are the same while messages are different. looks like copy-paste error.

Reply:
done